metricproviders/datadog/datadog.go

```diff
@@ -0,0 +1,166 @@
+package datadog
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/evaluate"
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+	templateutil "github.com/argoproj/argo-rollouts/utils/template"
+	log "github.com/sirupsen/logrus"
+	dd "github.com/zorkian/go-datadog-api"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// ProviderType indicates the provider is datadog
+	ProviderType = "Datadog"
+)
+
+// Provider contains all the required components to run a datadog query
+type Provider struct {
+	client *dd.Client
+	logCtx log.Entry
+}
+
+// Type incidates provider is a datadog provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+// Run queries DataDog for a metric
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+	newMeasurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	query, err := templateutil.ResolveArgs(metric.Provider.Datadog.Query, run.Spec.Args)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	// TODO (cabrinha) make from and to configurable
+	response, err := p.client.QueryMetrics(time.Now().Unix()-60, time.Now().Unix(), query)
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	newValue, newStatus, err := p.processResponse(metric, response[0])
+	if err != nil {
+		return metricutil.MarkMeasurementError(newMeasurement, err)
+	}
+
+	newMeasurement.Value = newValue
+	newMeasurement.Phase = newStatus
+	finishedTime := metav1.Now()
+	newMeasurement.FinishedAt = &finishedTime
+
+	return newMeasurement
+}
+
+// Resume should not be used with the datadog provider since all the work should occur in the Run method
+func (p *Provider) Resume(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Datadog provider should not execute the Resume method")
+	return measurement
+}
+
+// Terminate should not be used with the datadog provider since all the work should occur in the Run method
+func (p *Provider) Terminate(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, measurement v1alpha1.Measurement) v1alpha1.Measurement {
+	p.logCtx.Warn("Datadog provider should not execute the Terminate method")
+	return measurement
+}
+
+// GarbageCollect is a no-op for the datadog provider
+func (p *Provider) GarbageCollect(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric, limit int) error {
+	return nil
+}
+
+func (p *Provider) evaluateResult(result interface{}, metric v1alpha1.Metric) v1alpha1.AnalysisPhase {
+	successCondition := false
```

Comment:
We have a library for evaluateResults now at https://github.com/cabrinha/argo-rollouts/blob/master/utils/evaluate/evaluate.go#L13

Reply:
Alright, I deleted this whole func and used the logic I found in the prometheus code.